staging/src/k8s.io/client-go/discovery/discovery_client.go

```diff
@@ -29,9 +29,8 @@ import (
 	"sync"
 	"time"
 
-	//nolint:staticcheck // SA1019 Keep using module since it's still being maintained and the api of google.golang.org/protobuf/proto differs
-	"github.com/golang/protobuf/proto"
 	openapi_v2 "github.com/google/gnostic-models/openapiv2"
+	"google.golang.org/protobuf/proto"
```

Comment:
is the proto.Unmarshal call this switches to in the new library 100% equivalent?

Reply:
I think it is 100% equivalent.

The old implementation is:
```go
// Unmarshal parses a wire-format message in b and places the decoded results in m.
//
// Unmarshal resets m before starting to unmarshal, so any existing data in m is always
// removed. Use UnmarshalMerge to preserve and append to existing data.
func Unmarshal(b []byte, m Message) error {
	m.Reset()
	return UnmarshalMerge(b, m)
}

// UnmarshalMerge parses a wire-format message in b and places the decoded results in m.
func UnmarshalMerge(b []byte, m Message) error {
	mi := MessageV2(m)
	out, err := protoV2.UnmarshalOptions{
		AllowPartial: true,
		Merge:        true,
	}.UnmarshalState(protoiface.UnmarshalInput{
		Buf:     b,
		Message: mi.ProtoReflect(),
	})
	if err != nil {
		return err
	}
	if out.Flags&protoiface.UnmarshalInitialized > 0 {
		return nil
	}
	return checkRequiredNotSet(mi)
}

// UnmarshalState parses a wire-format message and places the result in m.
//
// This method permits fine-grained control over the unmarshaler.
// Most users should use [Unmarshal] instead.
func (o UnmarshalOptions) UnmarshalState(in protoiface.UnmarshalInput) (protoiface.UnmarshalOutput, error) {
	if o.RecursionLimit == 0 {
		o.RecursionLimit = protowire.DefaultRecursionLimit
	}
	return o.unmarshal(in.Buf, in.Message)
}
```
The new one is:
```go
// Unmarshal parses the wire-format message in b and places the result in m.
// The provided message must be mutable (e.g., a non-nil pointer to a message).
//
// See the [UnmarshalOptions] type if you need more control.
func Unmarshal(b []byte, m Message) error {
	_, err := UnmarshalOptions{RecursionLimit: protowire.DefaultRecursionLimit}.unmarshal(b, m.ProtoReflect())
	return err
}
```
So the old implementation is based on the new one, but:
- set `Merge: true`, but has explicit `m.Reset()` call. I assume this is necessary to support message V1, which we don't use now.
- set `AllowPartial: true` but has explicit `checkRequiredNotSet(mi)`. I assume this is necessary to return the error type defined in old package. But from doc: `RequiredNumbers is a list of required field numbers. In Proto3, it is always an empty list.`, this check is not relevant to proto3.

Reply:
[github.com/golang/protobuf/proto.Unmarshal](https://github.com/kubernetes/kubernetes/blob/95d71c464a6a5f81ae36bddc86972b3b2a1c40a5/vendor/github.com/golang/protobuf/proto/wire.go#L56) calls [`UnmarshalMerge`](https://github.com/kubernetes/kubernetes/blob/95d71c464a6a5f81ae36bddc86972b3b2a1c40a5/vendor/github.com/golang/protobuf/proto/wire.go#L62), which calls [`google.golang.org/protobuf/proto.UnmarshalState`](https://github.com/kubernetes/kubernetes/blob/95d71c464a6a5f81ae36bddc86972b3b2a1c40a5/vendor/google.golang.org/protobuf/proto/decode.go#L75) using the [`UnmarshalOptions`](https://github.com/kubernetes/kubernetes/blob/95d71c464a6a5f81ae36bddc86972b3b2a1c40a5/vendor/google.golang.org/protobuf/proto/decode.go#L24-L50):

```
AllowPartial: true,
Merge:        true,
```

Calling just [`google.golang.org/protobuf/proto.Unmarshal`](https://github.com/kubernetes/kubernetes/blob/95d71c464a6a5f81ae36bddc86972b3b2a1c40a5/vendor/google.golang.org/protobuf/proto/decode.go#L56) uses just the option: `RecursionLimit: protowire.DefaultRecursionLimit`.

- A `RecursionLimit` of `0` is euqal to `protowire.DefaultRecursionLimit`.
- The option `Merge` just means that we reset the input message before unmarshalling. Should have no effect in our case because the input is empty (`&openapi_v2.Document{}`)


The only difference I see now is `AllowPartial`, which will result in an error if there are any missing required fields. I can set it manually.

Reply:
> The only difference I see now is AllowPartial, which will result in an error if there are any missing required fields. I can set it manually.

@saschagrunert Please see above. I believe `AllowPartial` is not relevant to proto3. Unset fields will get zero value, and will not return error in either case.

Reply:
@huww98 makes sense.

Reply:
what does `MessageV2(...)` do to an instance of `&openapi_v2.Document{}`?

Reply:
For `openapi_v2.Document`, which is already a v2 message, it does nothing and just returns the input.
https://github.com/kubernetes/kubernetes/blob/70d3cc986aa8221cd1dfb1121852688902d3bf53/vendor/google.golang.org/protobuf/internal/impl/api_export.go#L114-L115

Reply:
if possible, we should remove this dependency on proto from the discovery_client https://github.com/kubernetes/kubernetes/issues/127888 , it will allow to build lighter clients, my local test shows about 8 M wins , from 50M to 42M if you build the clients without proto

Reply:
Could we stick to json here as well? The content type should be available, right?

```diff
diff --git a/staging/src/k8s.io/client-go/discovery/discovery_client.go b/staging/src/k8s.io/client-go/discovery/discovery_client.go
index 646820cbc2e..523656e607b 100644
--- a/staging/src/k8s.io/client-go/discovery/discovery_client.go
+++ b/staging/src/k8s.io/client-go/discovery/discovery_client.go
@@ -30,7 +30,6 @@ import (
 	"time"
 
 	openapi_v2 "github.com/google/gnostic-models/openapiv2"
-	"google.golang.org/protobuf/proto"
 
 	apidiscoveryv2 "k8s.io/api/apidiscovery/v2"
 	apidiscoveryv2beta1 "k8s.io/api/apidiscovery/v2beta1"
@@ -49,8 +48,6 @@ import (
 const (
 	// defaultRetries is the number of times a resource discovery is repeated if an api group disappears on the fly (e.g. CustomResourceDefinitions).
 	defaultRetries = 2
-	// protobuf mime type
-	openAPIV2mimePb = "application/com.github.proto-openapi.spec.v2@v1.0+protobuf"
 
 	// defaultTimeout is the maximum amount of time per request when no timeout has been set on a RESTClient.
 	// Defaults to 32s in order to have a distinguishable length of time, relative to other timeouts that exist.
@@ -668,12 +665,12 @@ func (d *DiscoveryClient) ServerVersion() (*version.Info, error) {
 
 // OpenAPISchema fetches the open api v2 schema using a rest client and parses the proto.
 func (d *DiscoveryClient) OpenAPISchema() (*openapi_v2.Document, error) {
-	data, err := d.restClient.Get().AbsPath("/openapi/v2").SetHeader("Accept", openAPIV2mimePb).Do(context.TODO()).Raw()
+	data, err := d.restClient.Get().AbsPath("/openapi/v2").Do(context.TODO()).Raw()
 	if err != nil {
 		return nil, err
 	}
 	document := &openapi_v2.Document{}
-	err = proto.Unmarshal(data, document)
+	err = json.Unmarshal(data, document)
 	if err != nil {
 		return nil, err
 	}
```

Reply:
From [protojson](https://pkg.go.dev/github.com/schattian/protobuf/encoding/protojson):
> ("encoding/json" package) does not operate correctly on protocol buffer messages.)